sql/execution_pipeline_impl.h

```diff
@@ -339,18 +339,47 @@ struct JoinElement: public PipelineElement {
         CrossJoinExecutor(const Bound * parent,
                           std::shared_ptr<ElementExecutor> root,
                           std::shared_ptr<ElementExecutor> left,
-                          std::shared_ptr<ElementExecutor> right);
+                          std::shared_ptr<ElementExecutor> right,
+                          size_t leftAdded,
+                          size_t rightAdded);
 
         const Bound * parent;
         std::shared_ptr<ElementExecutor> root, left, right;
+
+        bool wasOutput;
         
-        std::shared_ptr<PipelineResults> l,r;
+        std::shared_ptr<PipelineResults> l,r;     
+
+        const size_t leftAdded, rightAdded;   
             
         virtual std::shared_ptr<PipelineResults> take();
 
         void restart();
     };
 
+    struct FullCrossJoinExecutor: public ElementExecutor {
+        FullCrossJoinExecutor(const Bound * parent,
+                          std::shared_ptr<ElementExecutor> root,
+                          std::shared_ptr<ElementExecutor> left,
+                          std::shared_ptr<ElementExecutor> right,
+                          size_t leftAdded,
+                          size_t rightAdded);
+
+        const Bound * parent;
+        std::shared_ptr<ElementExecutor> root, left, right;
+        virtual std::shared_ptr<PipelineResults> take();
+
+        std::shared_ptr<PipelineResults> r;
+        typedef std::list<std::pair<std::shared_ptr<PipelineResults>, bool > > bufferType;
+        bufferType bufferedLeftValues;
+        bufferType::iterator l;
+        bool firstSpin;
+        bool wasOutput;
```

Comment:
It's not clear to me why we need `wasOutput` in addition to the flag in the buffer to tell if the row was outputted or not?